README.md
## WP Fragment Cache

### Description
**This plugin requires PHP version 5.3.6 or greater and can't be activated without a [persistent backend](https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Cache_Plugins) for the [WordPress Object Cache](https://codex.wordpress.org/Class_Reference/WP_Object_Cache).**

Improve website performance by caching individual page fragments (widgets, menus output and long loops).

Adds ability to cache particular sections of your site. This plugin is for people who can't use a full page cache or for sites with lots of pages and with frequent changes.
It can cache entire loops with the WP_Query(bypassing the database queries) or only the html output of the loop and if the query results has changed(comment added, post was edited, postmeta has changed ...) the cache will be invalidated and regenerated.

### Example usage:

```php
<?php if ( ! WP_Fragment_Cache::output( $wp_query_or_blockname, $duration ) ): ?>

  //content to be cached

  <?php WP_Fragment_Cache::store(); ?>
<?php endif; ?>
```

**Parameters:**

**$wp_query_or_blockname**
(mixed/string) (optional) a WP_Query(WP_Comment_Query) result or a string.
Default: the file and the line where the call was initiated(e.g. widgets/most-commented.php:18).

**$duration**
(int/string) (optional) Defines how many seconds to keep the cache for. If you pass 0 the cache will not expire. If you pass  the string "only_today" the cache will expire at the end of the current day. You can use [WordPress Time Constants](http://codex.wordpress.org/Transients_API#Using_Time_Constants).
Default: 86400 (one day).

### Advanced usage:
See the [wiki pages](https://github.com/mariuspass/wp-fragment-cache/wiki)

### Installation:
1. You should have a [persistent backend](https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Cache_Plugins) for the [WordPress Object Cache](https://codex.wordpress.org/Class_Reference/WP_Object_Cache) enabled. This plugin can't be activated without a persistent cache.
1. Download the plugin and unzip.
1. Upload them to `/wp-content/plugins/` directory on your WordPress installation.
1. Then activate the WP Fragment Cache plugin from Plugins page.
1. Edit your template files to include WP Fragment Cache. See the [Example usage](#example-usage)


CHANGELOG.md
## v1.0.4 (2014-06-18)
* PHP 5.3 compatibility fix.
* Make PHP version 5.3.6 minimal requirement.

## v1.0.3 (2014-04-28)
* Fix example usage code block.

## v1.0.2 (2014-04-26)
* Improve the description of the persistent cache requirement to avoid any confusion.

## v1.0.1 (2014-04-26)
* Add README and CHANGELOG.
* Add link for Advanced usage.

## v1.0.0 (2014-04-24)
* Initial release.


uninstall.php
<?php
/**
 * Fired when the plugin is uninstalled.
 *
 * @package   WP_Fragment_Cache
 * @author    Marius Dobre <mariuspass@gmail.com>
 * @license   GPL-2.0+
 * @link      https://github.com/mariuspass/WP-Fragment-Cache
 * @copyright 2014 Marius Dobre
 */

// If uninstall not called from WordPress, then exit
if ( ! defined( 'WP_UNINSTALL_PLUGIN' ) ) {
	exit;
}

if ( is_multisite() ) {
	global $wpdb;
	$blogs = $wpdb->get_results( "SELECT blog_id FROM {$wpdb->blogs}", ARRAY_A );
	uninstall_wp_cache_block_delete_options_and_files( false );
	if ( $blogs ) {
		foreach ( $blogs as $blog ) {
			switch_to_blog( $blog['blog_id'] );
			uninstall_wp_cache_block_delete_options_and_files();
			restore_current_blog();
		}
	}
} else {
	uninstall_wp_cache_block_delete_options_and_files();
}

/**
 * @param bool $optimize_tables
 */
function uninstall_wp_cache_block_delete_options_and_files( $optimize_tables = true ) {
	delete_option( 'wp_fragment_cache_is_enabled' );
	//info: remove custom file directory for main site
	$upload_dir = wp_upload_dir();
	$directory  = $upload_dir['basedir'] . DIRECTORY_SEPARATOR . 'CUSTOM_DIRECTORY_NAME' . DIRECTORY_SEPARATOR;
	if ( is_dir( $directory ) ) {
		foreach ( glob( $directory . '*.*' ) as $v ) {
			unlink( $v );
		}
		rmdir( $directory );
	}

	if ( $optimize_tables ) {
		//info: optimize tables
		$GLOBALS['wpdb']->query( 'OPTIMIZE TABLE `' . $GLOBALS['wpdb']->prefix . 'options`' );
	}
}

admin/views/admin.php
<?php
/**
 * Represents the view for the administration dashboard.
 *
 * This includes the header, options, and other information that should provide
 * The User Interface to the end user.
 *
 * @package   WP_Fragment_Cache
 * @author    Marius Dobre <mariuspass@gmail.com>
 * @license   GPL-2.0+
 * @link      https://github.com/mariuspass/WP-Fragment-Cache
 * @copyright 2014 Marius Dobre
 */
?>

<div class="wrap">

	<h2><?php echo esc_html( get_admin_page_title() ); ?></h2>

	<div class="postbox" style="width: 35%;margin-right: 10px;">
		<form method="post" action="options.php">
			<?php if ( $this->it_meets_the_requirements ): ?>
				<?php settings_fields( 'wp-fragment-cache-settings-group' ); ?>
				<?php do_settings_sections( 'wp-fragment-cache-settings-group' ); ?>
			<?php endif; ?>
			<?php $is_enabled_option = (bool) get_option( 'wp_fragment_cache_is_enabled' ); ?>
			<div class="inside">
				<table class="form-table">
					<tr valign="top">
						<th scope="row"><?php _e( 'Enabled' ); ?></th>
						<td>
							<p><label for="wp_fragment_cache_is_enabled_1">
									<input id="wp_fragment_cache_is_enabled_1" type="radio" value="1"
										   <?php checked( $is_enabled_option ); ?>
										   <?php disabled( ! $this->it_meets_the_requirements ); ?>
									       name="wp_fragment_cache_is_enabled">
									<?php _e( 'Yes' ) ?>
								</label>
							</p>

							<p>
								<label for="wp_fragment_cache_is_enabled_0">
									<input id="wp_fragment_cache_is_enabled_0" type="radio" value="0"
									       <?php checked( ! $is_enabled_option ); ?>
										   <?php disabled( ! $this->it_meets_the_requirements ); ?>
										   name="wp_fragment_cache_is_enabled">
									<?php _e( 'No' ) ?>
								</label>
							</p>
						</td>
					</tr>
				</table>
			</div>
			<div id="major-publishing-actions">
				<div id="publishing-action">
					<input type="submit" class="button-primary" <?php disabled( ! $this->it_meets_the_requirements ); ?>
						   value="<?php _e( 'Save' ) ?>">
				</div>
				<div class="clear"></div>
			</div>
		</form>
	</div>

	<div class="postbox" style="width: 35%;margin-right: 10px;">
		<form id="purgeall" action="" method="post" class="clearfix">
			<div class="inside">
				<table class="form-table">
					<tr valign="top">
						<th><?php _e( 'Purge All Cache' ); ?></th>

						<td>
							<?php if ( $this->it_meets_the_requirements && $is_enabled_option ): ?>
								<?php $purge_url  = add_query_arg( array( 'action' => 'purge' ) ); ?>
								<?php $nonced_url = wp_nonce_url( $purge_url, 'purge', 'wp_fragment_cache_nonce' ); ?>
								<a style="background: #DD3D36; border: 0;" href="<?php echo esc_url( $nonced_url ); ?>"
								   class="button button-primary button-large"><?php _e( 'Purge Cache' ); ?></a>
							<?php else : ?>
								<button style="background: #DD3D36 !important; border: 0;" disabled="disabled"
								   class="button button-primary button-large"><?php _e( 'Purge Cache' ); ?></button>
							<?php endif; ?>
						</td>
					</tr>
				</table>
			</div>
		</form>
	</div>

</div>


wp-fragment-cache.php
<?php
/**
 *
 * @package   WP_Fragment_Cache
 * @author    Marius Dobre <mariuspass@gmail.com>
 * @license   GPL-2.0+
 * @link      http://wordpress.org/extend/plugins/
 *
 * @wordpress-plugin
 * Plugin Name:       WP Fragment Cache
 * Plugin URI:        https://github.com/mariuspass/WP-Fragment-Cache
 * Description:       Boost your page performance by caching individual page fragments.
 * Version:           1.0.4
 * Author:            Marius Dobre
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * GitHub Plugin URI: https://github.com/mariuspass/WP-Fragment-Cache
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

/*----------------------------------------------------------------------------*
 * Public-Facing Functionality
 *----------------------------------------------------------------------------*/

require_once( plugin_dir_path( __FILE__ ) . 'public/class-wp-fragment-cache.php' );

/*
 * Register hooks that are fired when the plugin is activated or deactivated.
 * When the plugin is deleted, the uninstall.php file is loaded.
 */
register_activation_hook( __FILE__, array( 'WP_Fragment_Cache', 'activate' ) );
register_deactivation_hook( __FILE__, array( 'WP_Fragment_Cache', 'deactivate' ) );


add_action( 'plugins_loaded', array( 'WP_Fragment_Cache', 'get_instance' ) );

/*----------------------------------------------------------------------------*
 * Dashboard and Administrative Functionality
 *----------------------------------------------------------------------------*/

if ( is_admin() && ( ! defined( 'DOING_AJAX' ) || ! DOING_AJAX ) ) {
	require_once( plugin_dir_path( __FILE__ ) . 'admin/class-wp-fragment-cache-admin.php' );
	add_action( 'plugins_loaded', array( 'WP_Fragment_Cache_Admin', 'get_instance' ) );

}


admin/class-wp-fragment-cache-admin.php
<?php
/**
 * Plugin Name WP Fragment Cache
 *
 * @package   WP_Fragment_Cache_Admin
 * @author    Marius Dobre <mariuspass@gmail.com>
 * @license   GPL-2.0+
 * @link      https://github.com/mariuspass/WP-Fragment-Cache
 * @copyright Marius Dobre
 */

/**
 * @package WP_Fragment_Cache_Admin
 * @author  Marius Dobre <mariuspass@gmail.com>
 */
class WP_Fragment_Cache_Admin {

	/**
	 * Instance of this class.
	 *
	 * @since    1.0.0
	 *
	 * @var      object
	 */
	protected static $instance = null;

	/**
	 * Slug of the plugin screen.
	 *
	 * @since    1.0.0
	 *
	 * @var      string
	 */
	protected $plugin_screen_hook_suffix = null;

	/**
	 * It meets the requirements.
	 *
	 * @since    1.0.4
	 *
	 * @var      bool
	 */
	protected $it_meets_the_requirements;

	/**
	 * Initialize the plugin by adding a settings page and menu.
	 *
	 * @since     1.0.0
	 */
	private function __construct() {
		$this->plugin_slug               = WP_Fragment_Cache::get_instance()->get_plugin_slug();
		$this->it_meets_the_requirements = WP_Fragment_Cache::get_instance()->it_meets_the_requirements;

		// Add the options page and menu item.
		add_action( 'admin_menu', array( $this, 'add_plugin_admin_menu' ) );

		// Add an action link pointing to the options page.
		$plugin_basename = plugin_basename( plugin_dir_path( realpath( dirname( __FILE__ ) ) ) . $this->plugin_slug . '.php' );
		add_filter( 'plugin_action_links_' . $plugin_basename, array( $this, 'add_action_links' ) );

		if ( $this->it_meets_the_requirements ) {
			// Intercept option save
			add_filter( 'pre_update_option_wp_fragment_cache_is_enabled', array( $this, 'on_is_enabled_option_update' ), 10, 2 );
		}

	}

	/**
	 * Fired when is_enabled option is updated and purge the cache if is disabled.
	 *
	 * @since     1.0.0
	 *
	 * @param $new_value
	 * @param $old_value
	 *
	 * @return mixed
	 */
	public function on_is_enabled_option_update( $new_value, $old_value ) {
		if ( $new_value == 0 && $new_value !== $old_value ) {
			WP_Fragment_Cache::purge();
		}

		return $new_value;
	}

	/**
	 * Return an instance of this class.
	 *
	 * @since     1.0.0
	 *
	 * @return    object    A single instance of this class.
	 */
	public static function get_instance() {
		// If the single instance hasn't been set, set it now.
		if ( null == self::$instance ) {
			self::$instance = new self;
		}

		return self::$instance;
	}

	/**
	 * Register the administration menu for this plugin into the WordPress Dashboard menu.
	 *
	 * @since    1.0.0
	 */
	public function add_plugin_admin_menu() {

		$this->plugin_screen_hook_suffix = add_options_page(
			__( 'WP Fragment Cache Settings', $this->plugin_slug ),
			__( 'WP Fragment Cache', $this->plugin_slug ),
			'manage_options',
			$this->plugin_slug,
			array( $this, 'display_plugin_admin_page' )
		);

		add_action( 'load-'.$this->plugin_screen_hook_suffix, array( $this, 'on_load_plugin_admin_page' ) );

		if ( $this->it_meets_the_requirements ) {
			//call register settings function
			add_action( 'admin_init', array( $this, 'register_my_settings' ) );
		} else {
			add_action( 'admin_notices', array( $this, 'no_min_req_notice' ) );
		}
	}


	/**
	 * Fired when the plugin settings page is loaded and check if it need to purge the cache.
	 *
	 * @since    1.0.0
	 *
	 */
	public function on_load_plugin_admin_page() {
		if ( isset( $_GET['wp_fragment_cache_nonce'] ) && isset ( $_GET['action'] ) ) {
			if ( wp_verify_nonce( $_GET['wp_fragment_cache_nonce'], 'purge' ) ) {
				switch ( $_GET['action'] ) {
					case 'purge':
						WP_Fragment_Cache::purge();
						$redirect_url = remove_query_arg( 'settings-updated' );
						$redirect_url = add_query_arg( array( 'action' => 'done' ), $redirect_url );
						wp_redirect( $redirect_url );
						break;
					case 'done':
						if ( ! isset ( $_GET['settings-updated'] ) ) {
							add_action( 'admin_notices', array( $this, 'purge_done_notice' ) );
						}
						break;
				}
			}
		}
	}


	/**
	 * Show the no persistent cache notice.
	 *
	 * @since    1.0.0
	 */
	public function no_min_req_notice() {
		if ( ! wp_using_ext_object_cache() ) {
			echo '<div class="error">
				    <p>WP Fragment Cache cannot be enabled, because you don\'t have any
				        <a target="_blank" href="https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Caching">
						<strong>persistent cache</strong></a>.
				    </p>
				  </div>';
		}

		if ( version_compare( PHP_VERSION, WP_Fragment_Cache::PHP_MIN_VERSION, '<' ) ) {
			echo '<div class="error">
					  <p>WP Fragment Cache requires PHP version ' . WP_Fragment_Cache::PHP_MIN_VERSION . ' or greater.
				    </p>
				  </div>';
		}
	}

	/**
	 * Show the cache purge notice.
	 *
	 * @since    1.0.0
	 */
	public function purge_done_notice() {
		echo '<div class="updated"><p>Purge initiated</p></div>';
	}


	/**
	 * Register the plugin settings.
	 *
	 * @since    1.0.0
	 */
	public function register_my_settings() {
		register_setting( 'wp-fragment-cache-settings-group', 'wp_fragment_cache_is_enabled' );
	}

	/**
	 * Render the settings page for this plugin.
	 *
	 * @since    1.0.0
	 */
	public function display_plugin_admin_page() {

		include_once( 'views/admin.php' );
	}

	/**
	 * Add settings action link to the plugins page.
	 *
	 * @since    1.0.0
	 */
	public function add_action_links( $links ) {

		return array_merge(
			array(
				'settings' => '<a href="' . admin_url( 'options-general.php?page=' . $this->plugin_slug ) . '">' . __( 'Settings', $this->plugin_slug ) . '</a>'
			),
			$links
		);
	}

}


public/class-wp-fragment-cache.php
<?php
/**
 * Plugin Name WP Fragment Cache
 *
 * @package   WP_Fragment_Cache
 * @author    Marius Dobre <mariuspass@gmail.com>
 * @license   GPL-2.0+
 * @link      https://github.com/mariuspass/WP-Fragment-Cache
 * @copyright 2014 Marius Dobre
 */

/**
 * Plugin class. This class should ideally be used to work with the
 * public-facing side of the WordPress site.
 *
 * @package WP_Fragment_Cache
 * @author  Marius Dobre <mariuspass@gmail.com>
 */
class WP_Fragment_Cache {

	/**
	 * Plugin version, used for cache-busting of style and script file references.
	 *
	 * @since   1.0.0
	 *
	 * @var     string
	 */
	const VERSION = '1.0.4';

	/**
	 * Php minimum version requirement.
	 *
	 * @since   1.0.4
	 *
	 * @var     string
	 */
	const PHP_MIN_VERSION = '5.3.6';

	/**
	 *
	 * @since    1.0.0
	 *
	 * @var      string
	 */
	protected $plugin_slug = 'wp-fragment-cache';

	/**
	 * Instance of this class.
	 *
	 * @since    1.0.0
	 *
	 * @var      object
	 */
	protected static $instance = null;

	/**
	 * Key prefix.
	 *
	 * @since    1.0.0
	 *
	 * @var string
	 */
	private $cache_key_prefix;

	/**
	 * Key prefix.
	 *
	 * @since    1.0.0
	 *
	 * @var string
	 */
	private $theme_directory;

	/**
	 * Current wp_query or block name.
	 *
	 * @since    1.0.0
	 *
	 * @var string
	 */
	private $current_query_value;

	/**
	 * Is plugin enabled.
	 *
	 * @since   1.0.0
	 *
	 * @var bool
	 */
	public $is_enabled = false;

	/**
	 * It meets the requirements.
	 *
	 * @since   1.0.4
	 *
	 * @var bool
	 */
	public $it_meets_the_requirements = false;

	/**
	 * Cache duration.
	 *
	 * @since   1.0.0
	 *
	 * @var int
	 */
	public $duration = DAY_IN_SECONDS;

	/**
	 * Initialize the plugin by setting the options.
	 *
	 * @since     1.0.0
	 */
	private function __construct() {
		// Activate plugin when new blog is added
		add_action( 'wpmu_new_blog', array( $this, 'activate_new_site' ) );

		if ( wp_using_ext_object_cache() && version_compare( PHP_VERSION, self::PHP_MIN_VERSION, '>=' ) ) {
			$this->it_meets_the_requirements = true;

			if ( get_option( 'wp_fragment_cache_is_enabled' ) ) {
				$this->is_enabled      = true;
				$this->theme_directory = get_stylesheet_directory() . DIRECTORY_SEPARATOR;
			}
		}
	}

	/**
	 * Return the plugin slug.
	 *
	 * @since    1.0.0
	 *
	 * @return string  WP_Fragment_Cache slug variable.
	 */
	public function get_plugin_slug() {
		return $this->plugin_slug;
	}

	/**
	 * Return an instance of this class.
	 *
	 * @since     1.0.0
	 *
	 * @return    object    A single instance of this class.
	 */
	public static function get_instance() {

		// If the single instance hasn't been set, set it now.
		if ( null == self::$instance ) {
			self::$instance = new self;
		}

		return self::$instance;
	}

	/**
	 * Fired when the plugin is activated.
	 *
	 * @since    1.0.0
	 *
	 * @param    boolean $network_wide True if WPMU superadmin uses
	 *                                       "Network Activate" action, false if
	 *                                       WPMU is disabled or plugin is
	 *                                       activated on an individual blog.
	 */
	public static function activate( $network_wide ) {

		if ( function_exists( 'is_multisite' ) && is_multisite() ) {

			if ( $network_wide ) {

				// Get all blog ids
				$blog_ids = self::get_blog_ids();

				foreach ( $blog_ids as $blog_id ) {

					switch_to_blog( $blog_id );
					self::single_activate();
				}

				restore_current_blog();

			} else {
				self::single_activate();
			}
		} else {
			self::single_activate();
		}

	}

	/**
	 * Fired when the plugin is deactivated.
	 *
	 * @since    1.0.0
	 *
	 * @param    boolean $network_wide True if WPMU superadmin uses
	 *                                       "Network Deactivate" action, false if
	 *                                       WPMU is disabled or plugin is
	 *                                       deactivated on an individual blog.
	 */
	public static function deactivate( $network_wide ) {

		if ( function_exists( 'is_multisite' ) && is_multisite() ) {

			if ( $network_wide ) {

				// Get all blog ids
				$blog_ids = self::get_blog_ids();

				foreach ( $blog_ids as $blog_id ) {

					switch_to_blog( $blog_id );
					self::single_deactivate();

				}

				restore_current_blog();

			} else {
				self::single_deactivate();
			}
		} else {
			self::single_deactivate();
		}

	}

	/**
	 * Fired when a new site is activated with a WPMU environment.
	 *
	 * @since    1.0.0
	 *
	 * @param    int $blog_id ID of the new blog.
	 */
	public function activate_new_site( $blog_id ) {

		if ( 1 !== did_action( 'wpmu_new_blog' ) ) {
			return;
		}

		switch_to_blog( $blog_id );
		self::single_activate();
		restore_current_blog();

	}

	/**
	 * Get all blog ids of blogs in the current network that are:
	 * - not archived
	 * - not spam
	 * - not deleted
	 *
	 * @since    1.0.0
	 *
	 * @return   array|false    The blog ids, false if no matches.
	 */
	private static function get_blog_ids() {

		global $wpdb;

		// get an array of blog ids
		$sql = "SELECT blog_id FROM $wpdb->blogs WHERE archived = '0' AND spam = '0' AND deleted = '0'";

		return $wpdb->get_col( $sql );

	}

	/**
	 * Fired for each blog when the plugin is activated.
	 *
	 * @since    1.0.0
	 */
	private static function single_activate() {
		if ( wp_using_ext_object_cache() && version_compare( PHP_VERSION, self::PHP_MIN_VERSION, '>=' ) ) {
			$is_enabled = get_option( 'wp_fragment_cache_is_enabled', 'no_value_set' );
			if ( $is_enabled === 'no_value_set' ) {
				update_option( 'wp_fragment_cache_is_enabled', true );
			}
		} else {
			// set is_enabled false regardless of what was before
			update_option( 'wp_fragment_cache_is_enabled', false );
			add_action( 'admin_notices', array( 'WP_Fragment_Cache_Admin', 'no_min_req_notice' ) );
		}
	}

	/**
	 * Fired for each blog when the plugin is deactivated.
	 *
	 * @since    1.0.0
	 */
	private static function single_deactivate() {
		self::purge();
	}


	/**
	 * Set the key prefix for current fragment.
	 * The key will be the file and the line where the call was initiated.
	 * The file is relative to theme directory.
	 * e.g. widgets/most-commented.php:18 or inc/left-sidebar.php:55
	 *
	 * @since    1.0.0
	 */
	private function _set_cache_key_prefix() {
		// PHP 5.3 compatibility fix, limit parameter was added in php 5.4.0
		if ( version_compare( PHP_VERSION, '5.4.0' ) >= 0 ) {
			$backtraceArray = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS, 3 );
		} else {
			$backtraceArray = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
		}

		// Remove the theme root
		$backtraceArray[2]['file'] = str_replace( $this->theme_directory, '', $backtraceArray[2]['file'] );

		$this->cache_key_prefix = $backtraceArray[2]['file'] . ':' . $backtraceArray[2]['line'];
	}


	/**
	 * Echo the fragment if it's cached and is valid or turn on output buffering.
	 *
	 * @param mixed $wp_query_or_blockname
	 * @param null|int $duration
	 *
	 * @since    1.0.0
	 *
	 * @return bool
	 */
	private function _output( $wp_query_or_blockname, $duration ) {
		$this->_set_cache_key_prefix();

		if ( ! is_string( $wp_query_or_blockname ) ) { //$wp_query_or_blockname is an object or an array
			$this->current_query_value = sha1( serialize( $wp_query_or_blockname ) );
		} else {
			$this->cache_key_prefix   .= '_' . $wp_query_or_blockname;
			$this->current_query_value = '';
		}

		$cachedQueryValue = wp_cache_get( $this->cache_key_prefix . '_key', __CLASS__ );

		if ( $cachedQueryValue !== false ) { //the key exists
			if ( $cachedQueryValue === $this->current_query_value ) { //it's the same
				if ( WP_DEBUG ) {
					echo "<!-- WP_Fragment_Cache from Key '" . $this->cache_key_prefix . "' --> \n";
				}

				echo wp_cache_get( $this->cache_key_prefix . '_content', __CLASS__ );

				if ( WP_DEBUG ) {
					echo "<!-- WP_Fragment_Cache from Key '" . $this->cache_key_prefix . "' --> \n";
				}

				return true;
			} else { //has changed
				// delete the key and the content
				wp_cache_delete( $this->cache_key_prefix . '_key', __CLASS__ );
				wp_cache_delete( $this->cache_key_prefix . '_content', __CLASS__ );
			}
		}

		if ( $duration !== null ) {
			if ( $duration === 'only_today' ) {
				$now            = time();
				$this->duration = strtotime( date( 'Y-m-d' ) . ' 23:59:59', $now ) - $now;
			} else {
				$this->duration = intval( $duration );
			}
		}

		ob_start();

		return false;

	}

	/**
	 * Store the fragment in cache and add the key to All_Fragments_Keys_Array
	 *
	 * @since   1.0.0
	 */
	private function _store() {
		wp_cache_set( $this->cache_key_prefix . '_content', ob_get_contents(), __CLASS__, $this->duration );
		wp_cache_set( $this->cache_key_prefix . '_key', $this->current_query_value, __CLASS__, $this->duration );

		$all_Fragments_Keys_Array = wp_cache_get( 'All_Fragments_Keys_Array', __CLASS__ );
		if ( $all_Fragments_Keys_Array === false ) {
			$all_Fragments_Keys_Array = array();
		}

		$all_Fragments_Keys_Array[] = $this->cache_key_prefix;
		wp_cache_set( 'All_Fragments_Keys_Array', $all_Fragments_Keys_Array, __CLASS__ );

	}

	/**
	 * Purge all the fragments.
	 *
	 * @since   1.0.0
	 *
	 */
	static function purge() {
		$all_Fragments_Keys_Array = wp_cache_get( 'All_Fragments_Keys_Array', __CLASS__ );
		if ( $all_Fragments_Keys_Array !== false ) {
			foreach ( $all_Fragments_Keys_Array as $key ) {
				wp_cache_delete( $key . '_key', __CLASS__ );
				wp_cache_delete( $key . '_content', __CLASS__ );
			}
			wp_cache_delete( 'All_Fragments_Keys_Array', __CLASS__ );
		}
	}

	/**
	 * Static wrapper for _output()
	 * @see _output()
	 *
	 * @since    1.0.0
	 *
	 * @param mixed $wp_query_or_blockname
	 * @param null|int $duration in seconds or null for default duration(1 day)
	 * (you can use WP Time Constants
	 * @link http://codex.wordpress.org/Transients_API#Using_Time_Constants )
	 *
	 * @return bool
	 *
	 */
	static function output( $wp_query_or_blockname = '', $duration = null ) {
		$instance = WP_Fragment_Cache::get_instance();

		if ( $instance->is_enabled ) {
			return $instance->_output( $wp_query_or_blockname, $duration );
		}

		return false;
	}

	/**
	 * Static wrapper for _store()
	 * @see _store()
	 *
	 * @since    1.0.0
	 */
	static function store() {
		$instance = WP_Fragment_Cache::get_instance();

		if ( $instance->is_enabled ) {
			$instance->_store();
		}
	}

}
